In [1]:
import pandas as pd 
import numpy as np
import seaborn as sns
import scipy
from matplotlib import pyplot as plt

In [2]:
#loading the dataset
df = pd.read_csv("healthcare-dataset-stroke-data.csv")

df.head()

,id,gender,age,hypertension,heart_disease,ever_married,work_type,Residence_type,avg_glucose_level,bmi,smoking_status,stroke
0,9046,Male,67.0,0,1,Yes,Private,Urban,228.69,36.6,formerly smoked,1
1,51676,Female,61.0,0,0,Yes,Self-employed,Rural,202.21,NaN,never smoked,1
2,31112,Male,80.0,0,1,Yes,Private,Rural,105.92,32.5,never smoked,1
3,60182,Female,49.0,0,0,Yes,Private,Urban,171.23,34.4,smokes,1
4,1665,Female,79.0,1,0,Yes,Self-employed,Rural,174.12,24.0,never smoked,1


# Part 1: 10-Point Inspection
## Step 1: Shape

In [3]:
#understanding the size and structure of the dataset
shape = df.shape

print (f"This dataset has {shape[0]} rows and {shape[1]} columns.")

This dataset has 5110 rows and 12 columns.


### Findings:

How many rows (observations)? 5110 \
How many columns (features)? 12 \
What does each row represent in clinical terms? Each row represents one patient.

## Step 2: Column Names

In [10]:
#looking at the features of the dataset
print(df.columns.tolist())

['id', 'gender', 'age', 'hypertension', 'heart_disease', 'ever_married', 'work_type', 'Residence_type', 'avg_glucose_level', 'bmi', 'smoking_status', 'stroke']


### Findings: 

List all column names: id, gender, age, hypertension, heart_disease, ever_married, work_type, Residence_type, avg_glucose_level, bmi, smoking_status, stroke 

Which columns might need further research to understand? hypertension, average glucose level, bmi, stroke 

## Step 3: Data Types

In [5]:
#gathering informatiton about the data
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 5110 entries, 0 to 5109
Data columns (total 12 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   id                 5110 non-null   int64  
 1   gender             5110 non-null   str    
 2   age                5110 non-null   float64
 3   hypertension       5110 non-null   int64  
 4   heart_disease      5110 non-null   int64  
 5   ever_married       5110 non-null   str    
 6   work_type          5110 non-null   str    
 7   Residence_type     5110 non-null   str    
 8   avg_glucose_level  5110 non-null   float64
 9   bmi                4909 non-null   float64
 10  smoking_status     5110 non-null   str    
 11  stroke             5110 non-null   int64  
dtypes: float64(3), int64(4), str(5)
memory usage: 479.2 KB


### Findings: 
Which columns are numeric? 7 

Which columns are categorical (object/string)? 5

Are there any data types that seem incorrect? No

## Step 4: First Look

In [12]:
#looking at the first 10 rows of data
df.head(10)

,id,gender,age,hypertension,heart_disease,ever_married,work_type,Residence_type,avg_glucose_level,bmi,smoking_status,stroke
0,9046,Male,67.0,0,1,Yes,Private,Urban,228.69,36.6,formerly smoked,1
1,51676,Female,61.0,0,0,Yes,Self-employed,Rural,202.21,NaN,never smoked,1
2,31112,Male,80.0,0,1,Yes,Private,Rural,105.92,32.5,never smoked,1
3,60182,Female,49.0,0,0,Yes,Private,Urban,171.23,34.4,smokes,1
4,1665,Female,79.0,1,0,Yes,Self-employed,Rural,174.12,24.0,never smoked,1
5,56669,Male,81.0,0,0,Yes,Private,Urban,186.21,29.0,formerly smoked,1
6,53882,Male,74.0,1,1,Yes,Private,Rural,70.09,27.4,never smoked,1
7,10434,Female,69.0,0,0,No,Private,Urban,94.39,22.8,never smoked,1
8,27419,Female,59.0,0,0,Yes,Private,Rural,76.15,NaN,Unknown,1
9,60491,Female,78.0,0,0,Yes,Private,Urban,58.57,24.2,Unknown,1


### Findings: 
What do the actual values look like? The columns for each row describe the patient's basic demographic information (age, gender, etc.) as well as relevant health information (glucose level, bmi, history of heart disease). 

Do you notice anything unusual or unexpected? No

Are there any values that look like they might be placeholders for missing data? Yes, some BMI values are reported as null values (i.e. "NaN"). Additionally, some values for the smoking status column are reported as "unknown," which could indicate missing data. 

## Step 5: Last Look

In [15]:
#looking at the last 10 rows of data
df.tail(10)

,id,gender,age,hypertension,heart_disease,ever_married,work_type,Residence_type,avg_glucose_level,bmi,smoking_status,stroke
5100,68398,Male,82.0,1,0,Yes,Self-employed,Rural,71.97,28.3,never smoked,0
5101,36901,Female,45.0,0,0,Yes,Private,Urban,97.95,24.5,Unknown,0
5102,45010,Female,57.0,0,0,Yes,Private,Rural,77.93,21.7,never smoked,0
5103,22127,Female,18.0,0,0,No,Private,Urban,82.85,46.9,Unknown,0
5104,14180,Female,13.0,0,0,No,children,Rural,103.08,18.6,Unknown,0
5105,18234,Female,80.0,1,0,Yes,Private,Urban,83.75,NaN,never smoked,0
5106,44873,Female,81.0,0,0,Yes,Self-employed,Urban,125.20,40.0,never smoked,0
5107,19723,Female,35.0,0,0,Yes,Self-employed,Rural,82.99,30.6,never smoked,0
5108,37544,Male,51.0,0,0,Yes,Private,Rural,166.29,25.6,formerly smoked,0
5109,44679,Female,44.0,0,0,Yes,Govt_job,Urban,85.28,26.2,Unknown,0


### Findings: 

Does the data end cleanly? Yes

Are the last rows consistent with the first rows? Yes

## Step 6: Memory Usage

In [4]:
#checking the memory usage of the dataframe and converting it into megabytes
memory = df.memory_usage(deep=True).sum()/(1024**2)

print(f"Total Memory Usage: {memory: .2f} MB")

Total Memory Usage:  1.81 MB


### Findings: 
How much memory does the dataset use? 1.81 MB \
Is this a "small" or "large" dataset by data science standards? This is considered a small dataset. 

## Step 7: Missing Values

In [7]:
#converting any placeholders for missing/null data into detectable null data
df["smoking_status"] = df["smoking_status"].replace("Unknown", np.nan)

#identifying any missing and null values in the dataset
missing_count = df.isnull().sum()


percent_missing = missing_count * 100/len(df)

print("-----------------------------------------------")
print("Columns with missing or null values")
print("-----------------------------------------------")
print(missing_count)
print("-----------------------------------------------")
print("Percentage of each column that is missing data")
print("-----------------------------------------------")
print(percent_missing)
print("-----------------------------------------------")

-----------------------------------------------
Columns with missing or null values
-----------------------------------------------
id                      0
gender                  0
age                     0
hypertension            0
heart_disease           0
ever_married            0
work_type               0
Residence_type          0
avg_glucose_level       0
bmi                   201
smoking_status       1544
stroke                  0
dtype: int64
-----------------------------------------------
Percentage of each column that is missing data
-----------------------------------------------
id                    0.000000
gender                0.000000
age                   0.000000
hypertension          0.000000
heart_disease         0.000000
ever_married          0.000000
work_type             0.000000
Residence_type        0.000000
avg_glucose_level     0.000000
bmi                   3.933464
smoking_status       30.215264
stroke                0.000000
dtype: float64
-------------

### Findings: 
Which columns have missing values (according to .isnull())? The "bmi" and "smoking_status" columns have missing values.

What percentage of each column is missing? 3.93% of data in the "bmi" column is missing, and 30.2% in the "smoking_status" column is missing. 

⚠️ IMPORTANT: Are there any values that look like data but actually represent missing information? (Hint: Look carefully at the actual values in each column) Yes. Originally, missing values for the "smoking_status" column were reported as "Unknown." To be detected by pandas, these values were converted to NaN values. 

## Step 8: Duplicates

In [27]:
#identifying any duplicates in the dataset
num_duplicates = df.duplicated().sum()
id_duplicates = df.duplicated(subset = "id").sum()

print("-----------------------------------------------")
print(f"There are {num_duplicates} duplicates in this dataset.")
print(f"There are {id_duplicates} duplicates in the 'id' column.")
print("-----------------------------------------------")

-----------------------------------------------
There are 0 duplicates in this dataset.
There are 0 duplicates in the 'id' column.
-----------------------------------------------


### Findings: 
Are there any duplicate rows? No \
Are all patient IDs unique? Yes

## Step 9: Basic Statistics

In [32]:
#conducting basic descriptive statistics on the dataset
df.describe().round(3)

,id,age,hypertension,heart_disease,avg_glucose_level,bmi,stroke
count,5110.000,5110.000,5110.000,5110.000,5110.000,4909.000,5110.000
mean,36517.829,43.227,0.097,0.054,106.148,28.893,0.049
std,21161.722,22.613,0.297,0.226,45.284,7.854,0.215
min,67.000,0.080,0.000,0.000,55.120,10.300,0.000
25%,17741.250,25.000,0.000,0.000,77.245,23.500,0.000
50%,36932.000,45.000,0.000,0.000,91.885,28.100,0.000
75%,54682.000,61.000,0.000,0.000,114.090,33.100,0.000
max,72940.000,82.000,1.000,1.000,271.740,97.600,1.000


### Findings: 
What is the age range in the dataset? 0.08 years (approximately a month old) to 82 years \
What is the range of average glucose levels? 55.12 mg/dL to 271.74 mg/dL \
What is the range of BMI values? 10.3 to 97.6 \
Do any min/max values seem impossible or clinically unlikely? No 

## Step 10: Unique Counts

In [34]:
#looking at the number of unique values in each column
df.nunique()

id                   5110
gender                  3
age                   104
hypertension            2
heart_disease           2
ever_married            2
work_type               5
Residence_type          2
avg_glucose_level    3979
bmi                   418
smoking_status          4
stroke                  2
dtype: int64

### Findings: 
Which columns have very few unique values (likely categorical)? "gender", "hypertension", "heart_disease", "ever_married", "work_type", "residence_type", "smoking_status", and "stroke"

Which columns have many unique values (likely continuous or IDs)? "id", "age", "avg_glucose_level", and "bmi"

Does the number of unique IDs match the number of rows? Yes
